lib/trino-filesystem/src/main/java/io/trino/filesystem/local/LocalFileIterator.java

```diff
@@ -13,49 +13,73 @@
  */
 package io.trino.filesystem.local;
 
+import com.google.common.collect.ImmutableList;
 import io.trino.filesystem.FileEntry;
 import io.trino.filesystem.FileIterator;
 import io.trino.filesystem.Location;
 
 import java.io.IOException;
+import java.nio.file.FileVisitResult;
 import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
 import java.util.Iterator;
 import java.util.Optional;
-import java.util.stream.Stream;
 
-import static com.google.common.collect.ImmutableList.toImmutableList;
-import static io.trino.filesystem.local.LocalUtils.handleException;
 import static java.util.Collections.emptyIterator;
 import static java.util.Objects.requireNonNull;
 
 class LocalFileIterator
         implements FileIterator
 {
-    private final Path rootPath;
-    private final Iterator<Path> iterator;
+    private final Iterator<FileEntry> iterator;
 
     public LocalFileIterator(Location location, Path rootPath, Path path)
             throws IOException
     {
-        this.rootPath = requireNonNull(rootPath, "rootPath is null");
+        requireNonNull(rootPath, "rootPath is null");
         if (Files.isRegularFile(path)) {
             throw new IOException("Location is a file: " + location);
         }
         if (!Files.isDirectory(path)) {
             this.iterator = emptyIterator();
         }
         else {
-            try (Stream<Path> stream = Files.walk(path)) {
-                this.iterator = stream
-                        .filter(Files::isRegularFile)
-                        // materialize full list so stream can be closed
-                        .collect(toImmutableList())
-                        .iterator();
-            }
-            catch (IOException e) {
-                throw handleException(location, e);
-            }
+            ImmutableList.Builder<FileEntry> files = ImmutableList.builder();
+            Files.walkFileTree(path, new SimpleFileVisitor<>()
+            {
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attributes)
+                        throws IOException
+                {
+                    if (Files.isRegularFile(file)) {
+                        if (!file.startsWith(rootPath)) {
+                            throw new IOException("entry is not inside of filesystem root");
+                        }
+
+                        files.add(new FileEntry(
```

Comment:
The current approach may have the downside that between the time when the `FileEntry` gets created and when it gets visited it may be modified/deleted.
As long as this is used for testing I find this approach good enough.

Otherwise, pls consider investigating the strategy exposed in https://github.com/trinodb/trino/pull/20526#issuecomment-1946119026

Reply:
I can't understand https://github.com/trinodb/trino/pull/20526#issuecomment-1946119026. Can you elaborate on that?

Reply:
I'm implying to use a (potentially) simplified version of the algorithm https://github.com/openjdk/jdk/blob/18cea823a173e1b8b48d276daeca67b2a5cf3584/src/java.base/share/classes/java/nio/file/Files.java#L2798-L2848 based on `FileTreeWalker` which basically acts like an iterator.

Reply:
Thnank you for the hint @ebyhr .

Given that `FileTreeWalker` is a package-private tests and that `LocalFileIterator` is used only for test purposes - I consider that the downside of collecting early the files to list acceptable.